internal/core/application/trade_service.go

```diff
@@ -885,7 +922,24 @@ func previewFromFormula(
 	return price, previewAmount, nil
 }
 
-func isValidTradePrice(swapRequest *pb.SwapRequest, tradeType int, previewAmount uint64) bool {
+func isValidTradePrice(swapRequest *pb.SwapRequest, tradeType int, market *domain.Market) bool {
+	amount := swapRequest.AmountR
+	if tradeType == TradeSell {
+		amount = swapRequest.AmountP
+	}
+
+	_, previewAmount, _ := getPriceAndPreviewForMarket(
+		marketUnspents,
+		market,
+		tradeType,
+		amount,
+		market.BaseAsset,
```

Comment:
it shoudl switch based on the way of the tradeType

Reply:
the fact is that even depending on the trade type, it still can be possible that the trader fixed the quote asset amount to get the corresponding base asset one when getting the preview. For this reason, the daemon cannot be sure of the way it should verify prices, so I decided to always verify that the LBTC/ASSET price is valid.

Reply:
but since amount switches based on the trade type, also the asset should follow.

Reply:
the amount does not switch based on the trade type but on the asset passed to the MarketPrice. If I want to make a BUY passing an LBTC amount, the USDT one will be calculated and vice-versa.

For this reason, with the last commits, I added the validation for both the 2 predicted assets of a swap. I mean that if fixing the AssetP, the AmountR will be calculated based on the current price, and used to create a range in which the actual swap.AmountR must be included.

The same is done by fixing the AssetP though.

If both the 2 amounts of the swap are not included in the ranges, then the swap is invalid because of `bad pricing`. Walking the "wrong" way, which means fixing the amount that was actually calculated to predict the one that was actually fixed, does not grant to get back an amount that is enough close to the one of the swap. This depends essentially on the slippage % used to create the amount range.

For this reason, it's enough that one of the 2 ways result in a positive result to be sure that the amounts of the swap satisfy the current market price.